docs/adr/009_lns_affinity.md

```diff
@@ -0,0 +1,244 @@
+immediately (irrelevant but how does the Function get it for the deduplication decision?)
+
+Unclear what is the relationship between the twin and the cache on the function: in the cache we
+keep the framecounter down from the twin but not the fcnt up ❔
+
+- If both LNSs receive message B, LNS1 SHOULD reset the frame counter. LNS2 then
+call the Function first to reset the frame counter/clear the cache.
+  - The Function will update the device twin frame counter to 0 (it needs a DeviceClient that would
+    cause a temporary connection switch - to be investigated if this is a problem)
+  - Clear cache entry or update the entry with both frame counter down and up to 0.
+  - Returns the result to the LNS: whether it was the winning or losing one
+  - LNS reacts as described before
+
+Alternative would be to switch the order of the operations: first clear the cache and then (if we
+are the winning LNS) update the twin.
+
+#### Feature flag
+
+Unclear what is the exact use-case here ❔ We could still set the delay to 0.
+
+The stickiness feature can be disabled with a feature flag on the LNS configuration.
+
+- By default LNS implements sticky affinity. When the flag is set, LNS allow the connection
+  ping-pong to happen.
+- This flag is only checked if we are on the Drop deduplication strategy (as Mark and None do not
+  support stickiness anyway)
+
+#### Consequences / implementation
+
+- In memory storage is DevEui -> boolean.
+- Where to store locally on LNS❔ Can be on the
+
+  - LNS LoRaDeviceClient.ConnectionManager since all of the operations pass through it.
+  - LoRaDevice
+  - LoRaDeviceClient
+
+### Single point of connection handling on LoRaDevice
+
+TODO Atif's changes.
+
+## Other candidates considered
+
+### Using direct mode (not Edge hub)
+
+Less problematic than using Edge Hub but still has temporary connection stealing and can not be done
+due to the offline capabilities that the Edge Hub offers us.
+
+The [IoT Edge hub
+module](https://docs.microsoft.com/en-us/azure/iot-edge/iot-edge-runtime?view=iotedge-2020-11#iot-edge-hub)
+is responsible for communication, acting as a man-in-the-middle to IoT Hub. One of the major
+features it offers is [offline
+support](https://docs.microsoft.com/en-us/azure/iot-edge/offline-capabilities?view=iotedge-2020-11#how-it-works)
+that ensures messages are not lost and communication can still happen between child devices even
+when connectivity to Iot Hub is lost. When connectivity resumes it ensures the communication
+continues normally.
+
+
+### Parent-child gateways
+
+On deployments where multiple gateway servers are used, they need to have a parent 
+
+Child LNS A ____ | Parent Gateway using AMQP maintains a single connection to IoTHub | Child LNS B
+                ____|
+
+Not possible due to single parent limitation (not supporting roaming LNSs)
+
+## Questions
+
+❔ Should we investigate if creating a DeviceClient already steals the connection or we know this
```

Comment:
I don't think it matters. If we want to be sure, we dispose the client actively. We can't be sure about what happens in the SDK even if we verify that behavior now.

Reply:
True, the SDK behavior might change. And looking [at the current code where the SDK creates the DeviceClient](https://github.com/Azure/azure-iot-sdk-csharp/blob/2cf35f3ee541f3311cfcda3ce3b11290a9f88ae7/iothub/device/src/ClientFactory.cs#L485), they defer creating the right type of instance to later. Will remove this.